### PR TITLE
Rt 53 scan camera

### PR DESCRIPTION
**Pull request by rupertotorres1**
Opening custom camera after clicking the scan button (image looks stretched out vertically for now, though)

### CODE REVIEW

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        catch (e: CameraAccessException) {
+            e.printStackTrace()
+        }
+    }
+
+    private var mStateCallback: CameraDevice.StateCallback = object: CameraDevice.StateCallback() {
+        override fun onOpened(camera: CameraDevice?) {
+            Log.i(TAG, "onOpened")
+            mCameraDevice = camera
+            val texture = mTextureView.surfaceTexture
```
Comment (frostyshadows) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: You could do something like 
```
mTextureView.surfaceTexture?.let { texture ->
            texture.setDefaultBufferSize(mPreviewSize.width, mPreviewSize.height)
            val surface = Surface(texture)

            try {
                mPreviewBuilder = mCameraDevice!!.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)
            } catch (e: CameraAccessException) {
                e.printStackTrace()
            }

            mPreviewBuilder.addTarget(surface)

            try {
                mCameraDevice!!.createCaptureSession(arrayListOf(surface), mPreviewStateCallback, null)
            } catch (e: CameraAccessException) {
                e.printStackTrace()
            }
}
```
which is more Kotlin-y

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import android.hardware.camera2.CameraMetadata
+import android.hardware.camera2.CaptureRequest
+import android.os.Handler
+import android.view.Window
+import android.view.WindowManager
+
+
+class ScanCameraActivity : AppCompatActivity() {
+
+    companion object {
```
Comment (frostyshadows) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: can you put the companion object at the end of the class? That's been our convention so far for LaunchNote

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -4,4 +4,9 @@
     <string name="go_back_arrow">Go Back</string>
     <string name="toolbar_edit">Edit</string>
     <string name="title_activity_photo_info">Info</string>
+
+    <string name="title_activity_scan_camera">scanCameraActivity</string>
+    <string name="dummy_button">Dummy Button</string>
```
Comment (frostyshadows) [inline app/src/main/res/values/strings.xml]: Is this string ever used? If not, we don't really need it in strings.xml.

File: app/src/main/res/values/attrs.xml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<resources>
+
+    <!-- Declare custom theme attributes that allow changing which styles are
+         used for button bars depending on the API level.
+         ?android:attr/buttonBarStyle is new as of API 11 so this is
+         necessary to support previous API levels. -->
+    <declare-styleable name="ButtonBarContainerTheme">
```
Comment (frostyshadows) [inline app/src/main/res/values/attrs.xml]: Is this attribute actually used or are you planning to use it at some point?

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            val map = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
+            mPreviewSize = map.getOutputSizes(SurfaceTexture::class.java)[0]
+            manager.openCamera(cameraId, mStateCallback, null)
+        }
+        catch (e: CameraAccessException) {
+            e.printStackTrace()
+        }
+    }
+
+    private var mStateCallback: CameraDevice.StateCallback = object: CameraDevice.StateCallback() {
```
Comment (frostyshadows) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: maybe add a comment explaining what this does/how it's used.

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            Log.e(TAG, "onDisconnected");
+        }
+
+        override fun onError(camera: CameraDevice?, int: Int) {
+            //TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+            Log.e(TAG, "onError")
+        }
+    }
+
+    private val mPreviewStateCallback = object : CameraCaptureSession.StateCallback() {
```
Comment (frostyshadows) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: would also be good to have a comment saying what this callback is for.

**Review by frostyshadows — CHANGES_REQUESTED** [review]
It looks good, and doesn't actually get stretched out on my Nexus 5X emulator. 
Spend some time trying to have the camera be not stretched on your device as well, but if you get stuck on that, we can merge this PR first then make a new ticket for the stretching.
One bug: If I deny the app access to camera, I just see a black screen and have to hit the back button to get back to the app. It would be good if denying access to camera brings users back to the PhotoBrowser.

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            mPreviewBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO)
+
+            val backgroundThread = HandlerThread("CameraPreview")
+            backgroundThread.start()
+            val backgroundHandler = Handler(backgroundThread.looper)
+
+            try {
+                mPreviewSession.setRepeatingRequest(mPreviewBuilder.build(), null, backgroundHandler)
+            } catch (e: CameraAccessException) {
+                e.printStackTrace()
```
Comment (vpnda) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt:192]: I think we should display a Toast message here to notify the user.

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 1 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: done

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 2 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: done

File: app/src/main/res/values/strings.xml (hunk 3 above)
Comment (rupertotorres1) [reply app/src/main/res/values/strings.xml]: should I just delete all of these?

    <string name="title_activity_photo_info">Info</string>

    <string name="title_activity_scan_camera">scanCameraActivity</string>
    <string name="dummy_button">Dummy Button</string>
    <string name="dummy_content">DUMMY\nCONTENT</string>
    <string name="title_activity_test">test</string>

File: app/src/main/res/values/attrs.xml (hunk 4 above)
Comment (rupertotorres1) [reply app/src/main/res/values/attrs.xml]: I don't think so

File: app/src/main/res/values/strings.xml (hunk 3 above)
Comment (rupertotorres1) [reply app/src/main/res/values/strings.xml]: done (deleted dummy_button, dummy_content, and test)

File: app/src/main/res/values/attrs.xml (hunk 4 above)
Comment (rupertotorres1) [reply app/src/main/res/values/attrs.xml]: done

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+import android.os.HandlerThread
+import android.hardware.camera2.CameraMetadata
+import android.hardware.camera2.CaptureRequest
+import android.os.Handler
+import android.view.Window
+import android.view.WindowManager
+import com.example.ubclaunchpad.launchnote.photoBrowser.PhotoBrowserActivity
+
+
+class ScanCameraActivity : AppCompatActivity() {
```
Comment (vpnda) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: file name doesnt start with upper caps letter

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                    safeCameraOpen()
+                } else {
+                    // permission denied
+                    val browserIntent = Intent(this, PhotoBrowserActivity::class.java)
+                    startActivity(browserIntent)
+                }
+                return
+            }
+
+            else -> {}
```
Comment (vpnda) [inline app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: Toast msg or print to console stuff!!

**Review by vpnda — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 5 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: done

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 8 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: not sure why, it starts with upper case letter in my computer

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 7 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt:192]: notify about what? the exception?

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 9 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: done

File: app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt (hunk 6 above)
Comment (rupertotorres1) [reply app/src/main/java/com/example/ubclaunchpad/launchnote/scan/scanCameraActivity.kt]: done

**Review by frostyshadows — APPROVED** [review]

**Review by vpnda — APPROVED** [review]

**Pull request merged**